### ISSUE TITLE
Removing "tns:" namespace not correct in all cases (PR #506)

### ISSUE THREAD

**Issue by menssen** [vpulim/node-soap#537]
It appears to me that PR #506 broke the request generation for the API I am working with. Additionally, there seem to be issues with the way the ignoredNamespaces option is handled both before and after PR #506.

First, the issue I am having. The current version of the library generates this request:

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:tns="http://example.com/example/" xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <soap:Body>
        <ValidateSsoToken>
            <ssoToken>example</ssoToken>
            <ssoType>example</ssoType>
        </ValidateSsoToken>
    </soap:Body>
</soap:Envelope>
```

On this particular service, this fails with a "cannot find dispatch method" error. Before PR #506, the "tns:" namespace was included on the ValidateSsoToken element. Adding it back in makes the request succeed:

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:tns="http://example.com/example/" xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <soap:Body>
        <tns:ValidateSsoToken>
            <ssoToken>example</ssoToken>
            <ssoType>example</ssoType>
        </tns:ValidateSsoToken>
    </soap:Body>
</soap:Envelope>
```

So it seems that in some cases, removing the tns: prefix is not the correct behavior.

This has lead me down a deep hole of digging into issues with the ignoreNamespaces code. Here is a summary of what I have found:

**ONE**
It appears that PR #506 attempts to respect the ignoredNamespaces configuration option. Here is the conditional that was added to check if the namespace should be included or not, which will exclude the namespace if and only if it matches the new logic added by PR #506, AND is included in the old ignoredNamespaces array, which theoretically is configurable. (isNamespaceIgnored is the new logic.)

``` javascript
  if (... && (!isNamespaceIgnored || this.ignoredNamespaces.indexOf(namespace) === -1)) {
```

I'm not even sure that it is correct to use the old ignoredNamespaces array for this purpose, but assuming for a moment that that makes sense...

**TWO**
The ignoredNamespaces configuration option is ignored. In the above conditional, "this.ignoredNamespaces" is referenced, but when the options are processed, the final ignoredNamespaces are set into "this.options.ignoredNamespaces", NOT "this.ignoredNamespaces"

``` javascript
  var ignoredNamespaces = options ? options.ignoredNamespaces : null;

  if (ignoredNamespaces &&
      (Array.isArray(ignoredNamespaces.namespaces) || typeof ignoredNamespaces.namespaces === 'string')) {
    if (ignoredNamespaces.override) {
      this.options.ignoredNamespaces = ignoredNamespaces.namespaces;
    } else {
      this.options.ignoredNamespaces = this.ignoredNamespaces.concat(ignoredNamespaces.namespaces);
    }
  } else {
    this.options.ignoredNamespaces = this.ignoredNamespaces;
  }
```

this.ignoredNamespaces continues to be the default set of ignored namespaces, so the configuration options are ignored by this new code. Either the new code needs to be changed to read the this.options version, or the option processing code needs to update this.ignoredNamespaces

**THREE**
The test for the ignoredNamespaces option seems to be broken. It passes, but as far as I can tell it doesn't test anything. Specifically:

The ignored namespaces for the test are set in this file:

``` javascript
// test/request-response-samples/Dummy__should_ignore_custom_defined_namespaces/options.json
{
  "namespaces": ["someIgnoredNamespace"],
  "override": false
}
```

There are two problems:

a) Based on which files are passed to which parameters in the request-response test code, it looks like they should be set in `wsdl_options.json`, NOT `options.json`

b) The key of the ignoredNamespaces option is never provided, just the object that it should have as a value. So the configuration option is never actually set for this test. In other words, the file should probably be:

``` javascript
// test/request-response-samples/Dummy__should_ignore_custom_defined_namespaces/wsdl_options.json
{
  "ignoreNamespaces":
    "namespaces": ["someIgnoredNamespace"],
    "override": false
  }
}
```

**In summary**
1. There needs to be a way to override the behavior introduced in PR #506 
2. It looks like the intention was to use ignoreNamespaces for this, which I don't know if is the correct approach or not
3. The code to implement that intention is broken
4. The test for ignoreNamespaces is broken
5. The combination of 1-4 means that I don't know the right way to fix it.

I'm happy to work on this but need some direction.

**Comment by herom** [comment]
I'm sorry for the late reply, but I'm currently very short on time. I just wanted to comment here and show you that I'm aware of this issue but I'm not sure if I can look into it within the next few days (hopefully I'll have some time this weekend to look into this more closely).

In general, namespaces keep biting us big time and I guess that with every "fix" we'll raise another "issue" as there is no clear workflow for resolving namespaces properly. This implies that the process of parsing the `wsdl` is really fragile and needs to be improved/rewritten.

Otherwise the namespace handling will get more error prone with every PR which "just fixes" a current issue (reminds me of Scrat -->)
![Scrat Nut](http://tclhost.com/vcz3IIK.gif)

**Comment by nfedyk** [comment]
I ran into exactly the same issue. "tbs" is being removed by default inside wsdl.js. It makes sense to let user override following set of namespaces as well, as ***ignoredNamespaces***
```javascript
WSDL.prototype._ignoredSchemaNamespaces = ['tns', 'xs', 'xsd'];
```

**Comment by sommestad** [comment]
I believe I've got the same issue. Overriding `ignoredNamespaces` has no effect, since `this.ignoredNamespaces` is never updated in `wsdl.js` (only `this.options.ignoredNamespaces` is).

Log output after having provided an empty array for `ignoredNamespaces`:

```
this.ignoredNamespaces= [ 'tns', 'targetNamespace', 'typedNamespace' ]
this.options.ignoredNamespaces= []
```

The following change resolved the issue for me (only locally, obviously):
```js
// use this.options.ignoredNamespaces instead
if (prefixNamespace && ((qualified || first) || soapHeader) && (!isNamespaceIgnored || this.options.ignoredNamespaces.indexOf(namespace) === -1)) {
```

**Comment by csulok** [comment]
Today I ran into this issue as well. I'm working on the KNIME Enterprise Server's SOAP API ([sample WSDL](http://hastebin.com/elemifumiq.xml)) where it looks like I need to reset the ignoreNamespaces array to make sure the requests are sent as the server expects them, but ```tns``` couldn't be removed with:

```
var options = {
	ignoredNamespaces: {
		namespaces: [],
		override: true
	}
};
soap.createClient(url, options, ...)
```

I have a patch similar to @sommestad 's suggestion, I could send that in a pull request, but its testing, especially considering earlier comments on ```wsdl``` library being fragile is beyond me for now.

https://github.com/csulok/node-soap/commit/f4b77380f8f7f5e6dd28b7177d94727d74d8ba35

**Comment by CThuleHansen** [comment]
Any news on this one? I am having the same issue

**Comment by riaan53** [comment]
@sommestad Thank you it solved my issue as well.

**Comment by tehsenaus** [comment]
+1

Here is the workaround I used:
```js
require('soap/lib/wsdl').WSDL.prototype._ignoredSchemaNamespaces = ['xs', 'xsd'];
```

**Comment by CThuleHansen** [comment]
I fixed it by manually entering namespaces in my json

**Comment by riaan53** [comment]
@tehsenaus thx that is a better workaround!

**Comment by tobias-neubert** [comment]
See https://github.com/vpulim/node-soap/pull/662. That one fixes the problems described here. I totally agree that it makes no sense to let the developer manually declare namespaces to be ignored and than let some kind of automatism override that configuration. Hence I removed the shouldIgnoreNamespace method.
